tracer/src/Datadog.Trace/Util/SpanCharSplitter.cs

```diff
@@ -0,0 +1,85 @@
+// <copyright file="SpanCharSplitter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Datadog.Trace.Util;
+
+internal readonly ref struct SpanCharSplitter
+{
+    private readonly string _source;
+    private readonly char _separator;
+    private readonly int _count;
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public SpanCharSplitter(string source, char separator, int count)
+    {
+        _source = source;
+        _separator = separator;
+        _count = count;
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public SpanSplitEnumerator GetEnumerator() => new(_source, _separator, _count);
+
+    internal ref struct SpanSplitEnumerator
+    {
+        private readonly char _separator;
+        private readonly string _source;
+        private int _nextStartIndex = 0;
+        private int _count;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public SpanSplitEnumerator(string source, char separator, int count)
+        {
+            _source = source;
+            _separator = separator;
+            _count = count;
+        }
+
+        public SpanSplitValue Current { get; private set; }
+
+        public bool MoveNext()
+        {
+            if (_nextStartIndex > _source.Length)
+            {
+                return false;
+            }
+
+            var foundIndex = _source.IndexOf(_separator, _nextStartIndex);
+
+            var length = _count > 1 && foundIndex >= 0 ? foundIndex - _nextStartIndex : _source.Length - _nextStartIndex;
```

Comment:
I'm struggling to follow this 😅 This is because if we are limiting the number of values to find, we should specify the length as the _rest_ of the split?

Reply:
Exactly. I'm mimicking the behavior of `string.Split`. So if count is 1 OR there is no separator left, we return the rest of the split.